File: java/src/test/java/com/pqc/composite/CompositeSigTest.java
package com.pqc.composite;

import com.pqc.composite.CompositeSig.CompositeKeyPair;
import com.pqc.composite.CompositeSig.Scheme;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for composite signature schemes.
 */
class CompositeSigTest {

    @ParameterizedTest
    @EnumSource(Scheme.class)
    void testRoundtrip(Scheme scheme) {
        CompositeKeyPair kp = CompositeSig.keyGen(scheme);
        byte[] msg = ("Hello " + scheme.name()).getBytes();
        byte[] sig = CompositeSig.sign(kp, msg);
        assertTrue(CompositeSig.verify(scheme, kp.pk(), msg, sig),
            scheme.name() + ": valid signature must verify");
    }

    @ParameterizedTest
    @EnumSource(Scheme.class)
    void testWrongMessage(Scheme scheme) {
        CompositeKeyPair kp = CompositeSig.keyGen(scheme);
        byte[] msg = "Original".getBytes();
        byte[] sig = CompositeSig.sign(kp, msg);
        assertFalse(CompositeSig.verify(scheme, kp.pk(), "Tampered".getBytes(), sig),
            scheme.name() + ": wrong message must fail");
    }

    @ParameterizedTest
    @EnumSource(Scheme.class)
    void testTamperClassical(Scheme scheme) {
        CompositeKeyPair kp = CompositeSig.keyGen(scheme);
        byte[] msg = "Tamper classical".getBytes();
        byte[] sig = CompositeSig.sign(kp, msg);
        byte[] tampered = Arrays.copyOf(sig, sig.length);
        if (tampered.length > 4) {
            tampered[4] ^= (byte) 0xFF;
        }
        assertFalse(CompositeSig.verify(scheme, kp.pk(), msg, tampered),
            scheme.name() + ": tampered classical sig must fail");
    }

    @ParameterizedTest
    @EnumSource(Scheme.class)
    void testTamperPQ(Scheme scheme) {
        CompositeKeyPair kp = CompositeSig.keyGen(scheme);
        byte[] msg = "Tamper PQ".getBytes();
        byte[] sig = CompositeSig.sign(kp, msg);
        byte[] tampered = Arrays.copyOf(sig, sig.length);
        tampered[tampered.length - 1] ^= (byte) 0xFF;
        assertFalse(CompositeSig.verify(scheme, kp.pk(), msg, tampered),
            scheme.name() + ": tampered PQ sig must fail");
    }

    @Test
    void testMlDsa65Ed25519KeySizes() {
        Scheme scheme = Scheme.MLDSA65_ED25519;
        CompositeKeyPair kp = CompositeSig.keyGen(scheme);
        // pk = 32 (Ed25519) + 1952 (ML-DSA-65)
        assertEquals(32 + scheme.pqParams.pkSize, kp.pk().length,
            "composite pk size");
        // sk = 32 (Ed25519 seed) + 4032 (ML-DSA-65)
        assertEquals(32 + scheme.pqParams.skSize, kp.sk().length,
            "composite sk size");
    }

    @Test
    void testShortSigRejects() {
        assertFalse(CompositeSig.verify(
            Scheme.MLDSA65_ED25519, new byte[1984], "msg".getBytes(), new byte[2]));
    }
}

File: go/composite/composite_sig_test.go
package composite

import (
	"bytes"
	"testing"
)

func testRoundtrip(t *testing.T, scheme Scheme, name string) {
	t.Helper()
	kp := KeyGen(scheme)
	msg := []byte("Composite signature test: " + name)
	sig := Sign(kp, msg)
	if !Verify(scheme, kp.PK, msg, sig) {
		t.Fatalf("%s: valid signature did not verify", name)
	}
}

func testWrongMessage(t *testing.T, scheme Scheme, name string) {
	t.Helper()
	kp := KeyGen(scheme)
	msg := []byte("Original message " + name)
	sig := Sign(kp, msg)
	if Verify(scheme, kp.PK, []byte("Tampered message"), sig) {
		t.Fatalf("%s: wrong message should not verify", name)
	}
}

func testTamperClassical(t *testing.T, scheme Scheme, name string) {
	t.Helper()
	kp := KeyGen(scheme)
	msg := []byte("Tamper classical " + name)
	sig := Sign(kp, msg)
	tampered := make([]byte, len(sig))
	copy(tampered, sig)
	if len(tampered) > 4 {
		tampered[4] ^= 0xFF
	}
	if Verify(scheme, kp.PK, msg, tampered) {
		t.Fatalf("%s: tampered classical sig should not verify", name)
	}
}

func testTamperPQ(t *testing.T, scheme Scheme, name string) {
	t.Helper()
	kp := KeyGen(scheme)
	msg := []byte("Tamper PQ " + name)
	sig := Sign(kp, msg)
	tampered := make([]byte, len(sig))
	copy(tampered, sig)
	tampered[len(tampered)-1] ^= 0xFF
	if Verify(scheme, kp.PK, msg, tampered) {
		t.Fatalf("%s: tampered PQ sig should not verify", name)
	}
}

func testKeyFormat(t *testing.T, scheme Scheme, name string) {
	t.Helper()
	kp := KeyGen(scheme)
	params := pqParams(scheme)
	expectedPKSize := classicalPKSize(scheme) + params.PKSize
	expectedSKSize := classicalSKSize(scheme) + params.SKSize
	if len(kp.PK) != expectedPKSize {
		t.Fatalf("%s: pk size = %d, want %d", name, len(kp.PK), expectedPKSize)
	}
	if len(kp.SK) != expectedSKSize {
		t.Fatalf("%s: sk size = %d, want %d", name, len(kp.SK), expectedSKSize)
	}
}

func testParseSig(t *testing.T, scheme Scheme, name string) {
	t.Helper()
	kp := KeyGen(scheme)
	msg := []byte("Parse sig test " + name)
	sig := Sign(kp, msg)
	classical, pq, err := ParseSig(sig)
	if err != nil {
		t.Fatalf("%s: ParseSig error: %v", name, err)
	}
	if len(classical) == 0 || len(pq) == 0 {
		t.Fatalf("%s: empty component sig", name)
	}
	// Reconstruct and verify
	reconstructed := bytes.Join([][]byte{sig[:4], classical, pq}, nil)
	if !bytes.Equal(reconstructed, sig) {
		t.Fatalf("%s: reconstructed sig does not match", name)
	}
}

// ML-DSA-65 + Ed25519
func TestRoundtripMlDsa65Ed25519(t *testing.T)      { testRoundtrip(t, MlDsa65Ed25519, "ML-DSA-65+Ed25519") }
func TestWrongMsgMlDsa65Ed25519(t *testing.T)        { testWrongMessage(t, MlDsa65Ed25519, "ML-DSA-65+Ed25519") }
func TestTamperClassicalMlDsa65Ed25519(t *testing.T)  { testTamperClassical(t, MlDsa65Ed25519, "ML-DSA-65+Ed25519") }
func TestTamperPQMlDsa65Ed25519(t *testing.T)         { testTamperPQ(t, MlDsa65Ed25519, "ML-DSA-65+Ed25519") }
func TestKeyFormatMlDsa65Ed25519(t *testing.T)        { testKeyFormat(t, MlDsa65Ed25519, "ML-DSA-65+Ed25519") }
func TestParseSigMlDsa65Ed25519(t *testing.T)         { testParseSig(t, MlDsa65Ed25519, "ML-DSA-65+Ed25519") }

// ML-DSA-65 + ECDSA-P256
func TestRoundtripMlDsa65EcdsaP256(t *testing.T)     { testRoundtrip(t, MlDsa65EcdsaP256, "ML-DSA-65+ECDSA-P256") }
func TestWrongMsgMlDsa65EcdsaP256(t *testing.T)      { testWrongMessage(t, MlDsa65EcdsaP256, "ML-DSA-65+ECDSA-P256") }
func TestTamperClassicalMlDsa65EcdsaP256(t *testing.T) { testTamperClassical(t, MlDsa65EcdsaP256, "ML-DSA-65+ECDSA-P256") }
func TestTamperPQMlDsa65EcdsaP256(t *testing.T)       { testTamperPQ(t, MlDsa65EcdsaP256, "ML-DSA-65+ECDSA-P256") }
func TestKeyFormatMlDsa65EcdsaP256(t *testing.T)      { testKeyFormat(t, MlDsa65EcdsaP256, "ML-DSA-65+ECDSA-P256") }

// ML-DSA-87 + Ed25519
func TestRoundtripMlDsa87Ed25519(t *testing.T)       { testRoundtrip(t, MlDsa87Ed25519, "ML-DSA-87+Ed25519") }
func TestWrongMsgMlDsa87Ed25519(t *testing.T)         { testWrongMessage(t, MlDsa87Ed25519, "ML-DSA-87+Ed25519") }
func TestTamperClassicalMlDsa87Ed25519(t *testing.T)   { testTamperClassical(t, MlDsa87Ed25519, "ML-DSA-87+Ed25519") }
func TestTamperPQMlDsa87Ed25519(t *testing.T)          { testTamperPQ(t, MlDsa87Ed25519, "ML-DSA-87+Ed25519") }

// ML-DSA-44 + Ed25519
func TestRoundtripMlDsa44Ed25519(t *testing.T)       { testRoundtrip(t, MlDsa44Ed25519, "ML-DSA-44+Ed25519") }
func TestWrongMsgMlDsa44Ed25519(t *testing.T)         { testWrongMessage(t, MlDsa44Ed25519, "ML-DSA-44+Ed25519") }
func TestTamperClassicalMlDsa44Ed25519(t *testing.T)   { testTamperClassical(t, MlDsa44Ed25519, "ML-DSA-44+Ed25519") }
func TestTamperPQMlDsa44Ed25519(t *testing.T)          { testTamperPQ(t, MlDsa44Ed25519, "ML-DSA-44+Ed25519") }
